File: sdk/android/frak-sdk/src/main/kotlin/id/frak/sdk/Frak.kt
package id.frak.sdk

import android.app.Application
import android.content.Context
import id.frak.sdk.applink.AndroidAppLauncher
import id.frak.sdk.applink.DeepLinkObserver
import id.frak.sdk.config.SharedPreferencesStore
import id.frak.sdk.core.DeepLinkHandling
import id.frak.sdk.core.DefaultFrakClient
import id.frak.sdk.core.FrakConfig
import id.frak.sdk.core.FrakEnvironment
import id.frak.sdk.core.FrakError
import id.frak.sdk.core.FrakLogger
import id.frak.sdk.core.MainThreadDispatcher
import id.frak.sdk.core.TrackingConsent
import id.frak.sdk.core.defaultIoDispatcher
import id.frak.sdk.identity.AndroidKeystoreDeviceKeyStore
import id.frak.sdk.identity.AnonymousIdStore
import id.frak.sdk.sharing.FrakContext
import id.frak.sdk.sharing.SharingLinkBuilder
import id.frak.sdk.tracking.EventQueue
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.future.future
import java.io.File
import java.util.concurrent.CompletableFuture

/**
 * Entry point. Call [initialize] once from `Application.onCreate`, then use [client].
 *
 * Java callers use `Frak.getClient()` and the `*Async` twin of every suspending member.
 */
public object Frak {
    /**
     * Everything one [initialize] built, swapped as a unit. Three separate fields let a reader
     * observe half a teardown, and left the "these always move together" invariant to code review.
     */
    private class Session(
        val core: DefaultFrakClient,
        val client: FrakClient,
        /** Kept so [shutdown] can unregister it; otherwise re-initializing double-handles inbound deep links. */
        val observer: Pair<Application, Application.ActivityLifecycleCallbacks>?,
    )

    @Volatile
    private var session: Session? = null

    /** Non-blocking, does no I/O, never throws. Second call is a no-op; first config wins. */
    @JvmStatic
    public fun initialize(
        context: Context,
        config: FrakConfig,
    ) {
        val logger = FrakLogger(config.logLevel, config.logSink)
        if (session != null) {
            logger.warn("Frak.initialize was called more than once. The first configuration is kept.")
            return
        }
        synchronized(this) {
            if (session != null) return
            val effective = config.withPackageIdFrom(context)
            if (effective.merchantId == null && effective.packageId == null) {
                logger.error(
                    "FrakConfig has neither a merchantId nor a packageId. " +
                        "Every SDK call will fail with MerchantResolutionFailed.",
                )
            }
            (effective.env as? FrakEnvironment.Custom)?.rejectionReason?.let { reason ->
                logger.error("FrakEnvironment.Custom: $reason Requests will fail with FrakError.Network.")
            }
            // Shared by queue and client: two limitedParallelism(2) views would double the IO budget.
            val ioDispatcher = defaultIoDispatcher()
            // Separate prefs file from the config cache: a corrupt write must not take identity with it.
            val identityStore = SharedPreferencesStore(context, IDENTITY_FILE_NAME)
            // ONE instance, shared by the client and the identity store; two would drift on setTrackingEnabled.
            val consent =
                TrackingConsent(
                    store = identityStore,
                    configDefault = effective.trackingEnabled,
                    logger = logger,
                    ioDispatcher = ioDispatcher,
                )
            val newCore =
                DefaultFrakClient(
                    settings = effective,
                    store = SharedPreferencesStore(context),
                    // noBackupFilesDir: queued events must never be replayed from a backup/transfer.
                    queue =
                        EventQueue(
                            file = File(context.noBackupFilesDir, EVENT_QUEUE_FILE_NAME),
                            logger = logger,
                            ioDispatcher = ioDispatcher,
                        ),
                    identity =
                        AnonymousIdStore(
                            keyStore = AndroidKeystoreDeviceKeyStore(),
                            store = identityStore,
                            logger = logger,
                            merchantMarker = effective.merchantId ?: effective.packageId.orEmpty(),
                            consent = consent,
                            ioDispatcher = ioDispatcher,
                        ),
                    consent = consent,
                    launcher = AndroidAppLauncher(context),
                    logger = logger,
                    ioDispatcher = ioDispatcher,
                )
            // Built before the session is published and registered after it: a callback that fired
            // against a half-built session would drop the link it was handed.
            val registration = createDeepLinkObserver(context, effective, logger)
            session = Session(core = newCore, client = FrakClient(newCore), observer = registration)
            registration?.let { (application, callbacks) ->
                application.registerActivityLifecycleCallbacks(callbacks)
            }
            logger.info("Frak ${FrakSdkVersion.CURRENT} initialized.")
        }
    }

    /**
     * Tears the SDK down: cancels background coroutines, unregisters the deep-link observer, and
     * drops the client so [initialize] can run again. Not a privacy control — use
     * [FrakClient.setTrackingEnabled] for that. Idempotent; Java uses [shutdownAsync].
     */
    public suspend fun shutdown() {
        // State is read and cleared under the lock, then acted on outside it: `synchronized` must
        // never span a suspension point.
        val dying =
            synchronized(this) {
                val current = session
                session = null
                current
            }
        dying?.observer?.let { (application, callbacks) ->
            application.unregisterActivityLifecycleCallbacks(callbacks)
        }
        dying?.core?.shutdown()
    }

    /**
     * [shutdown] for Java. Runs on its own scope, not the client's, which `shutdown()` cancels.
     * Never `get()`/`join()` it on the main thread, and sequence a following [initialize] off the
     * future (`thenRun`) rather than beside it, or the new client races the old one's teardown.
     */
    @JvmStatic
    public fun shutdownAsync(): CompletableFuture<Void?> =
        teardownScope.future(MainThreadDispatcher) {
            shutdown()
            null
        }

    /** Not the client's scope, deliberately: see [shutdownAsync]. */
    private val teardownScope = CoroutineScope(SupervisorJob() + Dispatchers.Default)

    /** @throws FrakError.NotInitialized when [initialize] has not run. */
    @JvmStatic
    public val client: FrakClient
        get() = session?.client ?: throw FrakError.NotInitialized()

    /** Same as [client], but null instead of throwing. */
    @JvmStatic
    public val clientOrNull: FrakClient?
        get() = session?.client

    /** Whether [initialize] has run. For merchants guarding optional integrations. */
    @JvmStatic
    public val isInitialized: Boolean
        get() = session != null

    /** Pure and static; callable before [initialize]. Only decodes — does not track arrival. */
    @JvmStatic
    public fun parseReferralLink(url: String): FrakContext? = SharingLinkBuilder.parse(url)

    /**
     * Builds the observer without registering it, so [initialize] controls when it goes live.
     * Needs an `Application` to observe lifecycles; falls back to manual routing otherwise.
     */
    private fun createDeepLinkObserver(
        context: Context,
        config: FrakConfig,
        logger: FrakLogger,
    ): Pair<Application, Application.ActivityLifecycleCallbacks>? {
        if (config.deepLink != DeepLinkHandling.Automatic) return null
        val application = context.applicationContext as? Application
        if (application == null) {
            logger.error(
                "DeepLinkHandling.Automatic needs an Application context. " +
                    "Inbound referral links will be ignored; call handleReferralLink from your own router.",
            )
            return null
        }
        // Client owns the guard/tracking; this only reports that a link was seen. Reads the session
        // at call time, so a link arriving after [shutdown] reports nowhere instead of to a dead client.
        return application to DeepLinkObserver { url -> session?.core?.handleReferralLinkInBackground(url) }
    }

    private fun FrakConfig.withPackageIdFrom(context: Context): FrakConfig {
        if (merchantId != null || packageId != null) return this
        return withPackageId(context.packageName)
    }

    /**
     * Holds the merchant marker AND the consent decision, which is why this file is deliberately
     * left in Auto Backup: a withdrawal must survive a device transfer. The identity itself is not
     * in here — the keypair lives in `AndroidKeyStore` and cannot be backed up or transferred at
     * all — so there is nothing here to exclude. See `PRIVACY.md`.
     */
    private const val IDENTITY_FILE_NAME = "id.frak.sdk"

    private const val EVENT_QUEUE_FILE_NAME = "frak-events.jsonl"
}

File: sdk/ios/Sources/FrakSDK/Frak.swift
import Foundation

/// Entry point. Call `initialize(_:)` once, then use `client`.
///
/// ```swift
/// Frak.initialize(FrakConfig(merchantId: "...", metadata: FrakMetadata(name: "Acme")))
///
/// let reward = try await Frak.client.rewards.best(RewardRequest(targetInteraction: "purchase"))
/// ```
public enum Frak {
    private static let lock = NSLock()
    nonisolated(unsafe) private static var core: DefaultFrakClient?
    nonisolated(unsafe) private static var instance: FrakClient?

    // Never throws, and the only I/O is preparing the SDK's storage directory, which is memoised.
    // Second call is a no-op; the first configuration wins.
    public static func initialize(_ config: FrakConfig) {
        let logger = FrakLogger(level: config.logLevel, sink: config.logSink)
        let effective = config.withBundleIdFromMainBundle()
        let missingIdentity = effective.merchantId == nil && effective.bundleId == nil

        // Logging must not happen while `lock` is held (a sink reading Frak.client would
        // deadlock); the switch below emits every log line after the lock is released.
        enum Outcome {
            case alreadyInitialized
            case missingStore
            case missingIdentityStore
            case initialized
        }

        let outcome: Outcome = {
            lock.lock()
            defer { lock.unlock() }

            if instance != nil {
                return .alreadyInitialized
            }

            guard let store = UserDefaultsStore(),
                let consentStore = UserDefaultsStore(suiteName: UserDefaultsStore.consentSuiteName)
            else {
                return .missingStore
            }
            // Key material and the marker guarding it go to a backup-excluded file, never a
            // UserDefaults suite: a suite plist is carried to a restored device, which would
            // clone this installation's identity onto it.
            guard let identityStore = FileKeyValueStore.makeDefault(logger: logger) else {
                return .missingIdentityStore
            }

            // One instance, shared by the client and the identity store: two would memoise the
            // persisted decision independently and drift the moment setTrackingEnabled is called.
            // The consent suite, not the disposable config cache, and not the identity file: a
            // withdrawal is the one persisted value that must survive a restore.
            let consent = TrackingConsent(
                store: consentStore,
                configDefault: effective.trackingEnabled,
                logger: logger
            )
            let newCore = DefaultFrakClient(
                settings: effective,
                store: store,
                identity: AnonymousIdStore(
                    keyStore: PersistedDeviceKeyStore(store: identityStore, logger: logger),
                    store: identityStore,
                    logger: logger,
                    // App scope == merchant scope; regenerated if this ever changes.
                    merchantMarker: effective.merchantId ?? effective.bundleId ?? "",
                    consent: consent
                ),
                consent: consent,
                queue: EventQueue(fileURL: EventQueue.defaultFileURL(logger: logger), logger: logger),
                launcher: SystemAppLauncher(),
                logger: logger
            )
            core = newCore
            instance = FrakClient(core: newCore)
            return .initialized
        }()

        switch outcome {
        case .alreadyInitialized:
            logger.warn("Frak.initialize was called more than once. The first configuration is kept.")
        case .missingStore:
            if missingIdentity {
                logger.error(
                    "FrakConfig has neither a merchantId nor a bundleId. "
                        + "Every SDK call will fail with merchantResolutionFailed."
                )
            }
            logger.error("Frak could not open its UserDefaults suite. The SDK will not initialize.")
        case .missingIdentityStore:
            // Refuses rather than falling back to a location that is backed up or purgeable:
            // the first would clone this identity onto a restored device, the second would
            // report every purge as a brand-new user.
            logger.error(
                "Frak could not prepare its identity storage under Application Support. "
                    + "The SDK will not initialize."
            )
        case .initialized:
            if missingIdentity {
                logger.error(
                    "FrakConfig has neither a merchantId nor a bundleId. "
                        + "Every SDK call will fail with merchantResolutionFailed."
                )
            }
            if let reason = effective.env.customOriginRejectionReason {
                logger.error("FrakEnvironment.custom: \(reason) Requests will fail with FrakError.network.")
            }
            logger.info("Frak \(FrakSDKVersion.current) initialized.")
        }
    }

    public static var client: FrakClient {
        get throws {
            lock.lock()
            defer { lock.unlock() }
            guard let instance else { throw FrakError.notInitialized }
            return instance
        }
    }

    /// Same as `client`, but nil instead of throwing: for a call site that would just `try?`
    /// it anyway. Exists for parity with the Android surface, and for a call site that reads
    /// better without `try?`.
    public static var clientOrNull: FrakClient? {
        lock.lock()
        defer { lock.unlock() }
        return instance
    }

    // Pure/static: works before initialize(_:) has run. Decode-only — arrival tracking
    // and the self-referral guard live in FrakClient.appLink.handleReferral(_:).
    public static func parseReferralLink(_ url: String) -> FrakContext? {
        SharingLinkBuilder.parse(url)
    }

    public static func parseReferralLink(_ url: URL) -> FrakContext? {
        SharingLinkBuilder.parse(url.absoluteString)
    }

    public static var isInitialized: Bool {
        lock.lock()
        defer { lock.unlock() }
        return instance != nil
    }

    /// Tears the SDK down: cancels the background work it owns and drops the client so
    /// `initialize(_:)` can run again with a different `FrakConfig`.
    ///
    /// Not a privacy control — use `FrakClient.setTrackingEnabled(_:)` for that; shutting the
    /// SDK down neither records a consent decision nor erases anything, so a merchant who calls
    /// only this has stopped tracking for exactly as long as their process lives. Exists so a
    /// host can deterministically release the SDK, and so the facade is testable at all.
    ///
    /// Idempotent and safe before `initialize(_:)`.
    public static func shutdown() async {
        await takeForShutdown()?.shutdown()
    }

    /// Clears every singleton and hands back the client that was live, for `shutdown()` to await
    /// outside the lock.
    ///
    /// A separate synchronous function rather than a `lock()`/`unlock()` pair inside
    /// `shutdown()` itself: `NSLock.lock()`/`unlock()` are unavailable from an async context
    /// under Swift 6 strict concurrency, so the inline version failed the simulator build.
    /// Hoisting the critical section out also makes the "lock never spans a suspension point"
    /// rule structural: there is no `await` this function could reach.
    private static func takeForShutdown() -> DefaultFrakClient? {
        lock.lock()
        defer { lock.unlock() }
        let dying = core
        core = nil
        instance = nil
        return dying
    }

    /// Synchronous teardown for tests that cannot await. Drops the singletons WITHOUT cancelling
    /// the client's background work — prefer `shutdown()`, which does both.
    static func resetForTesting() {
        lock.lock()
        defer { lock.unlock() }
        core = nil
        instance = nil
    }
}

extension FrakConfig {
    fileprivate func withBundleIdFromMainBundle() -> FrakConfig {
        guard merchantId == nil, bundleId == nil, let bundleId = Bundle.main.bundleIdentifier else {
            return self
        }
        return withBundleId(bundleId)
    }
}
